ldelf/ta_elf.c

```diff
@@ -855,6 +862,87 @@ static void populate_segments(struct ta_elf *elf)
 	}
 }
 
+static void ta_elf_add_bti(struct ta_elf *elf)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct segment *seg = NULL;
+	uint32_t flags = LDELF_MAP_FLAG_EXECUTABLE | LDELF_MAP_FLAG_BTI;
+
+	TAILQ_FOREACH(seg, &elf->segs, link) {
+		vaddr_t va = elf->load_addr + seg->vaddr;
+
+		if (seg->flags & PF_X) {
+			res = sys_set_prot(va, seg->memsz, flags);
+			if (res)
+				err(res, "sys_set_prot");
+		}
+	}
+}
+
+static void parse_property_segment(struct ta_elf *elf)
+{
+	char *desc = NULL;
+	size_t align = elf->prop_align;
+	size_t desc_offset = 0;
+	size_t prop_offset = 0;
+	vaddr_t va = 0;
+	Elf_Note *note = NULL;
+	char *name = NULL;
+
+	if (!IS_ENABLED(CFG_TA_BTI) || !elf->prop_start)
+		return;
+
+	check_phdr_in_range(elf, PT_GNU_PROPERTY, elf->prop_start,
+			    elf->prop_memsz);
+
+	va = elf->load_addr + elf->prop_start;
+	note = (void *)va;
+	name = (char *)(note + 1);
+
+	if (elf->prop_memsz < sizeof(*note) + sizeof(ELF_NOTE_GNU))
+		return;
+
+	if (note->n_type != NT_GNU_PROPERTY_TYPE_0 ||
+	    note->n_namesz != sizeof(ELF_NOTE_GNU) ||
+	    memcmp(name, ELF_NOTE_GNU, sizeof(ELF_NOTE_GNU)))
+		return;
+
+	assert(IS_POWER_OF_TWO(align));
```

Comment:
This test should be unconditional. Asserts might not be enabled in all builds.

Reply:
Fixed